### ISSUE TITLE
Nodes remain in memory after removal with `del`

### ISSUE THREAD

**Issue by 8day** [vapoursynth/vapoursynth#399]
If you'll run the script below (don't mind globals) you'll see that Python process with loaded VS and FFMS2 consumes 27 MB, then memory consumption increases to 45 MB when we load clip, and after `del clip` it drops to 30 MB (+3 MB). When we fetch a few frames in same environment (after reloading clip), consumption increases to 67 MB, and after `del clip` it drops to 47 MB (+20 MB).

I've tried this ~20 times, on r41 and r43, and results are consistent. This happens even if you simply load clip w/o fetching frames (i.e., no cache bloat and dangling references to frames), as well as only one clip that is not referenced by any other filter.

This may be not a big issue for average 'procedural' scripts, but for interactive ones like VS editor running in same process/environment as VS/VPY, it is (Yuuno (?)).

I'm on Windows 7 64 bit with Python 3.6.5 and default MS binaries for R43.

```python
import os
import psutil


def get_process_memory():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss

def profile(func):
    def wrapper(*args, **kwargs):
        mem_before = get_process_memory()
        result = func(*args, **kwargs)
        mem_after = get_process_memory()
        print(f'{func.__name__}:\n\tbef: {mem_before:,},\n\taft: {mem_after:,},\n\tcon: {mem_after - mem_before:,}')
        return result
    return wrapper


@profile
def import_core_from_vs():
    global core
    from vapoursynth import core

@profile
def load_ffms2():
    core.std.LoadPlugin('ffms2.dll')

@profile
def load_clip():
    global clip
    clip = core.ffms2.Source('720x480.mkv')

@profile
def fetch_frames_from_clip():
    for i in range(1000): f = clip.get_frame(i)

@profile
def delete_clip():
    global clip
    del clip


print('\ncase #1: w/o fetching of frames')
import_core_from_vs()
load_ffms2()
load_clip()
delete_clip()

print('\ncase #2: with fetching of frames')
load_clip()
fetch_frames_from_clip()
delete_clip()
```

Here's results of execution on my machine:
```
case #1: w/o fetching of frames
import_core_from_vs:
	bef: 23,248,896,
	aft: 25,640,960,
	con: 2,392,064
load_ffms2:
	bef: 25,657,344,
	aft: 27,242,496,
	con: 1,585,152
load_clip:
	bef: 27,242,496,
	aft: 45,285,376,
	con: 18,042,880
delete_clip:
	bef: 45,285,376,
	aft: 30,019,584,
	con: -15,265,792

case #2: with fetching of frames
load_clip:
	bef: 30,023,680,
	aft: 45,486,080,
	con: 15,462,400
fetch_frames_from_clip:
	bef: 45,604,864,
	aft: 67,723,264,
	con: 22,118,400
delete_clip:
	bef: 67,723,264,
	aft: 47,173,632,
	con: -20,549,632
```

**Comment by myrsloik** [comment]
This is most likely due to python's memory handling, or that's how I explain it. Python's own memory allocator never releases memory back to the system (unlike VS). This means it kinda keeps going up for quite a bit until it reaches an almost steady state. Or that's how I explain it to myself, I've had VS debug builds where I keep track of every single referenced object/handle and that part of it is completely clean.

See if you can run it 10000 times without issues. It's an interesting problem but I think this is how good it will get.

**Comment by 8day** [comment]
Finally was able to test this... It seems that you are right.

Based on results of heap analysis of Python process with VMMap, Python allocates chunks of ~16 MB of memory (in some special cases; sometimes it uses much smaller blocks), then fills them with data, which later gets freed only partially (i.e., some small variable may still use it), hence the whole block cannot be freed. Essentially this is a [lacking] 'feature' of Python -- there's no 'memory compactor' and/or defragmenter.

BTW, here's some interesting numbers on how far such bloat can go (YUV frames at 704x480):
- 128.2 MB for Bilinear + CompatBGR32 + 470bg and 2000 read frames in 10 clips. Peak at ~320 MB;
- 132.4 MB for Bilinear + CompatBGR32 + 470bg and 1000 read frames in 20 clips. Peak at ~520 MB;
- 52.3 MB for Bilinear + CompatBGR32 + 470bg and 1 read frame in 10 clips. Peak at ~80 MB;
- 134.2 MB for Bilinear + CompatBGR32 + 470bg and 1 read frame in 50 clips. Peak at ~800 MB;
- 234.8 MB for Bilinear + CompatBGR32 + 470bg and 1 read frame in 100 clips. Peak at ~1.7 GB. Tried more clips, but something weird happened at 3.9 GB and Python crashed (some 32 bit limit?);
- 150.5 MB for FlipVertical and 1 read frame in 100 clips. Peak at ~1.45 GB;
- 96.2 MB for raw source and 1 read frame in 100 clips. Peak at ~1.45 GB;
- 160.8 MB for raw source and 1 read frame in 200 clips. Peak at ~3 GB.

In all these cases we should've had ~45 MB. Technically, it should be possible to have these in gigabytes, rather than few hundred megs, esp. if you work on HD/FHD/UHD. I suppose that the only way to change this is to write more agile/customizable memory manager and have it committed into CPython.

**Comment by myrsloik** [comment]
Yes, it's a mystery why CPython doesn't simply map things to malloc/free or made their memory allocator less fragmenting. Simply replacing it all with malloc/free from a modern C standard library is probably better. Use 64bit versions they're faster...

**Comment by dubhater** [comment]
> Technically, it should be possible to have these in gigabytes, rather than few hundred megs, esp. if you work on HD/FHD/UHD.

What does the resolution have to do with how much memory is not freed by Python? Even if Python doesn't free the memory it allocated for a VideoFrame Python object, it should still "destroy" the VideoFrame object, thus triggering a call to VSAPI::freeFrame, no?

**Comment by myrsloik** [comment]
Note how it's "peak" usage. That definitely sounds like it's before freeing anything...

**Comment by 8day** [comment]
> What does the resolution have to do with how much memory is not freed by Python? Even if Python doesn't free the memory it allocated for a VideoFrame Python object, it should still "destroy" the VideoFrame object, thus triggering a call to VSAPI::freeFrame, no?

You are absolutely right. That was a stupid mistake. Frame resolution plays no role in this. It's just that I'm trying to fix another leakage, and after loading HD video bloat increases much more noticeably than with SD. But that doesn't seem to be caused by VS, most likely it's caused by Qt's `QImage`.

> Note how it's "peak" usage. That definitely sounds like it's before freeing anything...

Sorry for the confusion with that "peak" usage ("Peak Working Set" from Windows Task Manager). It's so high because I used list comprehension to generate clips:
```python
clips = [core.ffms2.Source('704x480.mkv') for _ in range(10)]
for clip in clips:
    for i in range(1000): f = clip.get_frame(i)
```
BTW, the mentioned crash was caused by lack of free memory, so there's nothing wrong with that either.

I should not have posted those stats... My point was that "fragmentation" may have significant/noticeable impact on consumed memory and that in some complicated/artificial cases has a *potential* to grow to hundreds of megs, or even gigs. It's like with concatenation of strings with `+`/`__add__()`: most people won't notice O(n^2), but for some only `str.join()` will be usable.

tl;dr: there's nothing wrong with VapourSynth and this issue may be closed, though it's best to add some note to VS docs so that people knew that this is perfectly normal behavior.

**Comment by 8day** [comment]
Judging by the fact that you didn't close this issue, I assume you expected that there's more to it?



*First of all, to avoid any confusion I feel that I must state explicitly the obvious: programming is just a hobby of mine and I have almost no clue about how OSs work. I.e., all that will follow has extremely high possibility to be false.*

After some trial and error I have noticed that all filters cause leaking, and it has nothing to do with Qt, memory fragmentation (it's there, but it has comparably negligible effect), or Python's custom memory manager. This is based on the assumption that (1) stack trace shows the code that allocated memory (I may have misinterpreted what allocates what, etc.) and (2) Python has no power over memory allocated by `vscore.cpp/MemoryUse::allocBuffer()` (it may be that Python delays some commands in some way, etc.).

I came to such conclusion after analyzing stack trace, which in the majority of heavy cases ended with `vscore.cpp/MemoryUse::allocBuffer()` (line #264 in R43; there were calls to other functions, but those are from `ucrt.dll` and `ntdll.dll`) and `VapourSynthPluginInit + 0x4537` from FFMS2 (I have no debug symbols for it to provide more detailed info). Also, I noticed that all pages tend to be filled with data and there's barely any fragmentation. What's weird, is that 704x480 has almost same memory footprint for when clip was loaded and unloaded, which makes me wonder what influences this: dimensions, different compression, or something else.

Considering that memory consumption grows, and then *somewhat* stabilizes, yet when clip is deleted it doesn't get freed *completely*, it feels like it's something related to caching (though that partial freeing makes things WTFish...). It seems that such increases in memory consumption happen when new clips are created and have a few frames fetched from them, which may be another proof of cache-related issue (caching is tied to `VideoNode` (?)). Another proof of this may be that 704x480 leaks much less than 1280x720 (I think SD leaks *only* when you do some filtering), and latter leaks much less than 1920x1080, i.e. it's the frames that are being leaked. This is the only explanation I have for why memory consumption doesn't grow to infinity/is bounded, yet allocated blocks remain in memory after deletion of clips.

As for *some* FFMS2-allocated blocks, I can't explain why they remain on the heap: this may be related to VS, or may be FFMS2 own leak. Maybe they contain 'current frame'...?



Here's a video with me showing what happens on the heap: https://youtu.be/z2G-Cm-0Gwk It's with subs, because I have horrible pronunciation. I think you may find it useful to see if I'm right or wrong. And here's Python script that I used to collect snapshots:
```python
import os, psutil, time
import vapoursynth as vs

def report_consumption(desc, process = psutil.Process(os.getpid())):
    print(f'{process.memory_info().rss:,} bytes -', desc)

vs.core.std.LoadPlugin(r'c:\ffms2\x64\ffms2.dll')
report_consumption('loaded VS, FFMS2 and stats modules')
time.sleep(1)

paths = ('704x480.mkv', '1280x720.mkv', '1920x1080.mkv')
for path in paths:
	clip = vs.core.ffms2.Source(path)
	# clip = vs.core.resize.Bilinear(clip=clip, format=vs.COMPATBGR32, matrix_in_s='470bg')
	# clip = vs.core.resize.Bilinear(clip=clip, format=vs.GRAY8, matrix_in_s='470bg')
	# clip = vs.core.resize.Bilinear(clip=clip, format=vs.YUV444P8, matrix_in_s='470bg')
	# clip = vs.core.std.FlipVertical(clip=clip)
	# clip = vs.core.std.FlipVertical(clip=vs.core.std.FlipVertical(clip=clip))
	# clip = vs.core.misc.AverageFrames(clips=[clip[0:1000], clip[5:1005], clip[10:1010]], weights=[.5,.3,.2])
	report_consumption(f'loaded clip "{path}"')
	time.sleep(1)

	for i in range(1000):
		f = clip.get_frame(i)
	report_consumption('fetched a few frames')
	time.sleep(1)

	del f
	report_consumption('deleted "f"')
	time.sleep(1)

	del clip
	report_consumption('deleted "clip"')
	time.sleep(1)
```
`time.sleep(1)` is needed to allow VMMap to make snapshots, which it does with minimum interval of 1 second. BTW, it may seem that sleeping may prevent freeing of resources, but if you'll split this script into three separate ones and then run each of them in IDLE, you'll see that memory consumption remains effectively the same.

And... well... another set of stats...:
```
28,602,368 bytes - loaded VS, FFMS2 and stats modules

54,845,440 bytes - loaded clip "704x480.mkv"
79,265,792 bytes - fetched a few frames
79,265,792 bytes - deleted "f"
47,599,616 bytes - deleted "clip"

77,799,424 bytes - loaded clip "1280x720.mkv"
129,900,544 bytes - fetched a few frames
129,900,544 bytes - deleted "f"
89,571,328 bytes - deleted "clip"

144,146,432 bytes - loaded clip "1920x1080.mkv"
218,079,232 bytes - fetched a few frames
218,079,232 bytes - deleted "f"
152,932,352 bytes - deleted "clip"
```

P.S. In video and in here it's said that SD is 704x480, but in fact it's 720x480.

**Comment by sekrit-twc** [comment]
Starting with VapourSynth R44, the Windows build has an internal memory management system for frame buffers. This memory management will allocate up to the specified "max_cache_size". This memory is freed when the VapourSynth core is destroyed. The reason you do not see this happening in your use case is because the core is a singleton in each script.

**Comment by 8day** [comment]
I've read on Yuuno PyPI page that:

> From VapourSynth R44 onwards, we can reset core objects.

Does it mean this can be done through VPY, or only C++ API?

**Comment by myrsloik** [comment]
I have no idea what this reset is actually, you need to find out what they're doing

**Comment by 8day** [comment]
OK... So is there a way to destroy core w/o destroying Python process?

**Comment by myrsloik** [comment]
No, not really

**Comment by stuxcrystal** [comment]
@8day I think it was worded badly: Resetting core objects actually means opening a new VPyScriptEnvironment thus creating a new core object. This way it behaves more like VSEdit than an actual script.

**Comment by 8day** [comment]
@stuxcrystal So, you are essentially referring to https://github.com/vapoursynth/vapoursynth/pull/389 (?):

```python
env = vpy_current_environment()
# sometime later
with env:
  # Do stuff inside this env.
```
If so, then how do I use that environment? Do I `eval()` VPY script within `env` context and then all created clips are magically freed once I exit it?

**Comment by stuxcrystal** [comment]
Well. Not really. It returns a wrapper around the currently used environment of whatever software you are using.

It has a very narrow use-case where you have gotten access to the VPY-API from Python in any form and 
now need more control over the currently open script environment. (For example here: https://git.encode.moe/irrational-encoding-wizardry/yuuno/blob/master/yuuno_ipython/ipy_vs/vsscript.py#L124)

On https://git.encode.moe/irrational-encoding-wizardry/yuuno/blob/master/yuuno_ipython/ipy_vs/vsscript.py#L58 I implement a feature I call "Isolated-Core":
* It opens a VPyScriptEnvironment (L59),
* switches to the new environment (L62),
* `evals` some code inside said environment (L64),
* and then just frees the VPyScriptEnvironment (L68)
